File: cpp/059_SpiralMatrixII.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        int count = n*n;
        vector<vector<int>> ret(n, vector<int>(n,0));
        int left = 0;
        int right = n-1;
        int up = 0;
        int down = n-1;
        int num = 1;
        while(1){
            //to right
            if(num > count) break;
            for(int i = left; i <= right; ++i){
                ret[up][i] = num++;
            }
            up++;
            //to down
            if(num > count) break;
            for(int i = up; i <= down; ++i){
                ret[i][right] = num++;
            }
            right--;
            //to left
            if(num > count) break;
            for(int i = right; i >= left; --i){
                ret[down][i] = num++;
            }
            down--;
            //to up
            if(num > count) break;
            for(int i = down; i >= up; --i){
                ret[i][left] = num++;
            }
            left++;
        }
        return ret;
    }
};

File: js/059_SpiralMatrixII.js
/**
 * @param {number} n
 * @return {number[][]}
 */
var generateMatrix = function(n) {
    var count = n*n;
    var ret = [];
    for(var p = 0; p < n; ++p){
        var t = [];
        for(var q = 0; q < n; ++q){
            t.push(0);
        }
        ret.push(t);
    }
    
    var left = 0, right = n-1;
    var up = 0, down = n-1;
    var num = 1;
    while(1){
        //to right
        if(num > count) break;
        for(var l = left; l <= right; ++l){
            ret[up][l] = num++;
        }
        up++;
        //to down
        if(num > count) break;
        for(var u = up; u <= down; ++u){
            ret[u][right] = num++;
        }
        right--;
        //to left
        if(num > count) break;
        for(var r = right; r >= left; --r){
            ret[down][r] = num++;
        }
        down--;
        //to up
        if(num > count) break;
        for(var d = down; d >= up; --d){
            ret[d][left] = num++;
        }
        left++;        
    }
    return ret;
};